src/coreclr/src/vm/dllimport.cpp

```diff
@@ -6532,476 +6473,603 @@ INT_PTR NDirect::GetNativeLibraryExport(NATIVE_LIBRARY_HANDLE handle, LPCWSTR sy
             }
 
-            hmod = LocalLoadLibraryHelper(currLibNameVariation, flags, pErrorTracker);
+            hmod = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlags, pErrorTracker);
             if (hmod != NULL)
             {
                 return hmod;
             }
         }
-        else if ((callingAssembly != nullptr) && searchAssemblyDirectory)
+
+        // This may be an assembly name
+        // Format is "fileName, assemblyDisplayName"
+        MAKE_UTF8PTR_FROMWIDE(szLibName, wszLibName);
+        char *szComma = strchr(szLibName, ',');
+        if (szComma)
         {
-            hmod = LoadFromPInvokeAssemblyDirectory(callingAssembly, currLibNameVariation, loadWithAlteredPathFlags | dllImportSearchPathFlags, pErrorTracker);
-            if (hmod != NULL)
+            *szComma = '\0';
+            // Trim white spaces
+            while (COMCharacter::nativeIsWhiteSpace(*(++szComma)));
+
+            AssemblySpec spec;
+            if (SUCCEEDED(spec.Init(szComma)))
             {
-                return hmod;
+                // Need to perform case insensitive hashing.
+                SString moduleName(SString::Utf8, szLibName);
+                moduleName.LowerCase();
+
+                StackScratchBuffer buffer;
+                szLibName = (LPSTR)moduleName.GetUTF8(buffer);
+
+                Assembly *pAssembly = spec.LoadAssembly(FILE_LOADED);
+                Module *pModule = pAssembly->FindModuleByName(szLibName);
+
+                hmod = LocalLoadLibraryHelper(pModule->GetPath(), loadWithAlteredPathFlags | dllImportSearchPathFlags, pErrorTracker);
             }
         }
 
-        hmod = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlags, pErrorTracker);
-        if (hmod != NULL)
+        return hmod;
+    }
+
+    NATIVE_LIBRARY_HANDLE LoadLibraryModuleBySearch(NDirectMethodDesc *pMD, LoadLibErrorTracker *pErrorTracker, PCWSTR wszLibName)
+    {
+        STANDARD_VM_CONTRACT;
+
+        BOOL searchAssemblyDirectory;
+        DWORD dllImportSearchPathFlags;
+
+        GetDllImportSearchPathFlags(pMD, &dllImportSearchPathFlags, &searchAssemblyDirectory);
+
+        Assembly *pAssembly = pMD->GetMethodTable()->GetAssembly();
+        return LoadLibraryModuleBySearch(pAssembly, searchAssemblyDirectory, dllImportSearchPathFlags, pErrorTracker, wszLibName);
+    }
+
+    // This thread local and UseExtensionPoints class are used to guard against infinite recursion.
+    // The managed APIs (NativeLibarary.Load/TryLoad) that invoke managed extension points can also
+    // be called from within the extension points themselves. This class tracks whether or not the
+    // runtime is in a state where it should invoke those extension points (i.e. whether or not it
+    // is already in a load triggered by the managed APIs.
+    thread_local bool t_shouldInvokeExtensionPoints = true;
+    class UseExtensionPoints
+    {
+    public:
+        UseExtensionPoints()
```

Comment:
I think that this recursion block should rather be the AssemblyLoadContext - to disallow recursite call on the same AssemblyContext and the same name.

Also, if you block it in C#, it should work for both CoreCLR and Mono.

Reply:
I was trying to find where I could just have it in one place, so I had put it here since the `DllImportResolver` delegate is not tied to the ALC.

Reply:
Although `NativeLibrary.LoadLibraryCallbackStub` is already shared and can get at the ALC.

If the tracking goes on in the `AssemblyLoadContext`, I'd need to add some shared part that conditionally calls `LoadUnmanagedDll` based on the tracking - then that tracking could cover all three extension points.

Reply:
I am not sure whether we want to make this fix to touch DllImportResolver. It is not required to address the problem described in the issue, and I can only think about problems that it would introduce.

Would it make sense to do this fix for the ALC only, and keep DllImportResolver on the existing plan (ie keep calling it for DllImports only and not for NativeLibrary.Load)?

Reply:
I think it would make the API behaviour a bit less intuitive/obvious to me? Saying this API that takes an `Assembly` behaves as if `DllImport` was used in the given `Assembly` is an easy line to draw. But saying this API that takes an `Assembly` will not use the per-assembly callback for the given `Assembly`, but will use the extension points of the `AssemblyLoadContext` in which the given `Assembly` is loaded becomes less clear-cut.

Reply:
I am looking at this from the intended use of the APIs:

- `DllImportResolver` enables the assembly to resolve native libraries for its own DllImports. It is expected to be overriden by the assembly for itself only, and you can have just one of these registered per assembly. I cannot think about a scenario where it would be desirable for the DllImportResolver handler to be called for `NativeLibrary.Load` calls. If the assembly wants to call its own DllImportResolver handler, it can just do it directly.

- `AssemblyLoadContext` enables resolving native libraries for all assemblies loaded into it, without the assemblies itself being aware of it.

Reply:
If the starting premise is that the `NativeLoad.Load` that takes an `Assembly` should behave as if `DllImport` was used in the given assembly, shouldn't it do exactly that? Regardless of the expected use of the other APIs that can modify the behaviour for that assembly?

Reply:
> `NativeLoad.Load` that takes an Assembly should behave as if DllImport was used in the given assembly

That may have been over-simplification.

Reply:
Keeping the DllImportResolver as is also reduces breaking potential of the fix. E.g. somebody may have DllImportResolver that starts with:
```
if (name != "mydll") 
    throw new Exception();
```